nestauk/asf_heat_pump_suitability
# 68 Add grid capacity feature

---

# Description

This pull request introduces a new feature to calculate the grid capacity for heat pump installations per LSOA. The main changes include:

1. Implementation of data processing functions for each Distribution Network Operator (DNO): ENW, NPg, SPEN, SSEN, UKPN, and WPD.
2. Creation of a unified substation dataset combining data from all DNOs.
3. Development of functions to distribute substation headroom to LSOAs and calculate headroom per household.
4. Implementation of a heat pump suitability assessment function that calculates the percentage of households in each LSOA that could potentially install heat pumps based on available grid capacity.
5. Integration of all components into a main `calculate_grid_capacity()` function.

Fixes #68 

# Instructions for Reviewer

In order to test the code in this PR, you need to:

1. Run the main script containing the `calculate_grid_capacity()` function.
2. Check the output DataFrame for correct column names and data types, especially the 'heatpump_installation_percentage' column.
3. Check the incorporation of this output in the wider pipeline

Please pay special attention to:

1. The data processing functions for each DNO (ENW, NPg, SPEN, SSEN, UKPN, WPD) to ensure they correctly handle the specific data formats and structures for each operator.
2. The `distribute_substation_headroom()` function, which allocates substation capacity to LSOAs based on household count. Verify that this allocation is performed correctly and fairly.
4. The `assess_heatpump_suitability()` function, particularly the calculation of 'max_heatpumps' and 'heatpump_installation_percentage'. Ensure these calculations are accurate and properly consider the power requirements of heat pumps.
5. The main `calculate_grid_capacity()` function to confirm it correctly integrates all components and produces a coherent final dataset.
6. Error handling and edge cases, such as LSOAs with very low household counts or unusually high/low headroom values.

The expected output should be a DataFrame containing LSOA-level data, including the percentage of households that could install heat pumps in each LSOA. This percentage should never exceed 100% and should accurately reflect the relationship between available grid capacity and household count.

# Checklist:
